app/controllers/api/v2/registration_controller.rb

```diff
@@ -0,0 +1,58 @@
+module Api
+  module V2
+    class RegistrationController < V2::BaseController
+      include Api::Version2
+      include Foreman::Controller::Registration
+      include Foreman::Controller::Parameters::Host
+
+      rescue_from ActiveRecord::RecordNotFound do |error|
+        logger.info "#{error.message} (#{error.class})"
+        not_found(error.message)
+      end
+
+      api :GET, '/register', N_('Render Global registration template')
+      param :organization_id, :number, desc: N_("ID of the Organization to register the host in.")
+      param :location_id, :number, desc: N_("ID of the Location to register the host in.")
+      param :hostgroup_id, :number, desc: N_("ID of the Host group to register the host in.")
+      param :operatingsystem_id, :number, desc: N_("ID of the Operating System to register the host in.")
+      def global
+        find_global_registration
+
+        unless @provisioning_template
+          not_found _('Global Registration Template with name %s defined via default_global_registration_item Setting not found, please configure the existing template name first') % Setting[:default_global_registration_item]
+          return
+        end
+
+        render plain: @provisioning_template.render(variables: @global_registration_vars).html_safe
+      end
+
+      api :POST, "/register", N_("Find or create a host and render the Host registration template")
+      param_group :host, Api::V2::HostsController
+      def host
+        begin
+          prepare_host
+          @template = @host.registration_template
```

Comment:
When I run the registration but the template is not correctly assigned, the host is created however the registration fails with  'Unable to find registration template for host tower.example.tst running CentOS 7.8.2003'

Could we split prepare_host to two stages and only save the host if we are able to find the template? That makes it more atomic

Reply:
> Could we split prepare_host to two stages and only save the host if we are able to find the template? That makes it more atomic

But this will if host is new - not created yet, only initialized. You will always get error `[Foreman::Exception]: Must provide an operating system` raised by `@host.registration_template`. You must save `@host` before finding host template.